[KubeRay][Docs] Simplify KubeRay upgrade steps and always upgrade CRDs first (#44826)
ray-project/ray: doc/source/cluster/kubernetes/user-guides/upgrade-guide.md

## Upgrade KubeRay

Upgrading the KubeRay version is the best strategy if you have any issues with KubeRay.
Due to reliability and security implication of webhooks, KubeRay doesn't support a conversion webhook to convert v1alpha1 to v1 APIs.

To upgrade the KubeRay version, it's highly recommended to follow these steps in order:
1. Upgrade the CRD manifest, containing new fields added to the v1 CRDs.
2. Upgrade the kuberay-operator image to the new version
3. Verify successful upgrade

> Here's an example when upgrading KubeRay from v1.0.0 to v1.1.0:

Author: Added an example for v0.6.0 -> v1.0.0 for consistency with the old docs

```
# upgrade the CRD to v1.1.0
# NOTE: here we use kubectl because Helm does not support lifecycle management of CRDs
# See the Helm documentation for more details: https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations

> $ kubectl apply --server-side --force-conflicts -k "github.com/ray-project/kuberay/ray-operator/config/crd?ref=v1.1.0"

Author: The `--force-conflict` is needed here to resolve server-side conflicts with helm:
```
$ kubectl apply --server-side -k "github.com/ray-project/kuberay/ray-operator/config/crd?ref=v1.1.0"
Apply failed with 2 conflicts: conflicts with "helm" using apiextensions.k8s.io/v1:
- .metadata.annotations.controller-gen.kubebuilder.io/version
- .spec.versions
Please review the fields above--they currently have other managers. Here
are the ways you can resolve this warning:
* If you intend to manage all of these fields, please re-run the apply
  command with the `--force-conflicts` flag.
* If you do not intend to manage all of the fields, please edit your
  manifest to remove references to the fields that should keep their
  current managers.
* You may co-own fields by updating your manifest to match the existing
  value; in this case, you'll become the manager if the other manager(s)
  stop managing the field (remove it from their configuration).
See https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts
```

I think it's fine to force resolve these conflicts because Helm is not managing CRDs after initial install anyways.

## Upgrade KubeRay

Upgrading the KubeRay version is the best strategy if you have any issues with KubeRay.

> Due to reliability and security implication of webhooks, KubeRay doesn't support a conversion webhook to convert v1alpha1 to v1 APIs.

Reviewer: ```suggestion
Due to reliability and security implications of webhooks, KubeRay doesn't support a conversion webhook to convert v1alpha1 to v1 APIs.
```

> To upgrade the KubeRay version, it's highly recommended to follow these steps in order:

Reviewer: ```suggestion
To upgrade the KubeRay version, follow these steps in order:
```

1. Upgrade the CRD manifest, containing new fields added to the v1 CRDs.

> 2. Upgrade the kuberay-operator image to the new version

Reviewer: ```suggestion
2. Upgrade the kuberay-operator image to the new version.
```

> 3. Verify successful upgrade

Reviewer: ```suggestion
3. Verify the success of the upgrade.
```

> Here's an example when upgrading KubeRay from v1.0.0 to v1.1.0:

Reviewer: ```suggestion
The following is an example of upgrading KubeRay from v1.0.0 to v1.1.0:
```

```

> # upgrade the CRD to v1.1.0

Reviewer: ```suggestion
# Upgrade the CRD to v1.1.0.
```

> # NOTE: here we use kubectl because Helm does not support lifecycle management of CRDs

Reviewer: ```suggestion
# Note: This example uses kubectl because Helm doesn't support lifecycle management of CRDs.
```

# See the Helm documentation for more details: https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations

> $ kubectl replace -k "github.com/ray-project/kuberay/ray-operator/config/crd?ref=v1.1.0"

Reviewer 2: What's the reason for using `kubectl replace`? Is it to avoid the size limit of `kubectl.kubernetes.io/last-applied-configuration`?

Author: Yes. Using `kubectl apply --server-side` also works around the size limit, but I think `kubectl replace` here is fine because we do not expect multiple writers to the CRD anyways and the manifest in release should be the source of truth overwriting any previous updates by other writers

Author: With `kubectl replace` we need to be careful not to use the `--force` command because that will delete the CRD and recreate it. But I think it's fine here as long as we don't recommend it

Author: Using `kubectl apply --server-side --force-conflicts -k "github.com/ray-project/kuberay/ray-operator/config/crd?ref=v1.0.0"` also works well, so I'm fine with putting either one in the docs

Reviewer 2: Got it. Thank you for the explanation!

> # Upgrade kuberay-operator to v1.1.0. This step does not upgrade the CRDs.

Reviewer: ```suggestion
# Upgrade kuberay-operator to v1.1.0. This step doesn't upgrade the CRDs.
```

$ helm upgrade kuberay-operator kuberay/kuberay-operator --version v1.1.0

> # Install a RayCluster using the v1.1.0 helm chart to verify successful upgrade

Reviewer: ```suggestion
# Install a RayCluster using the v1.1.0 helm chart to verify the success of the upgrade.
```

$ helm install raycluster kuberay/ray-cluster --version 1.1.0
```

> Here's an example when upgrading KubeRay from v0.6.0 to v1.0.0:

Reviewer: ```suggestion
The following is an example of upgrading KubeRay from v0.6.0 to v1.0.0:
```

```

> # upgrade the CRD to v1.0.0

Reviewer: ```suggestion
# Upgrade the CRD to v1.0.0.
```

> # NOTE: here we use kubectl because Helm does not support lifecycle management of CRDs

Reviewer: ```suggestion
# Note: This example uses kubectl because Helm doesn't support lifecycle management of CRDs.
```

# See the Helm documentation for more details: https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations

> $ kubectl replace -k "github.com/ray-project/kuberay/ray-operator/config/crd?ref=v1.0.0"

Reviewer 3: FYI this step causes disruptions for any running `RayService` resources until you apply the `RayService` config again. See this slack thread for more details: https://ray-distributed.slack.com/archives/C02GFQ82JPM/p1701370178891199

Author: Is that specific to  upgrading the CRD first only or are you referring to the entire v0.6.0 to v1.0.0 upgrade? As far as I'm aware the 0.6.0 -> 1.0.0 is known to be disruptive and breaking changes like that are expected but I'm not super familiar with RayService changes from 0.6.0 and 1.0.0 yet to know

Reviewer 2: cc @andrewsykim do you plan to address this comment in this PR? If you want to do it in a follow-up PR, I am comfortable merging this PR.

Reviewer 3: If you're planning to add the 0.6.0 -> 1.0.0 guidance in a separate PR, I'd suggest changing the example here to show upgrade from 1.0.0 -> 1.1.0 or something like that otherwise it could be misleading

Reviewer 2: @smit-kiri are you interested in opening a PR to share your experience after this PR got merged?

Author: The old guidance suggests deleting the CRD, so I think what we have in this PR is still better 

I don't have enough experience with 0.6.0 to 1.0.0 upgrade issues right now so it'll have to be a follow up PR if we add one.

Reviewer 3: @kevin85421 Yeah I can open a PR!

Reviewer 2: Thanks, @smit-kiri!

> # Upgrade kuberay-operator to v1.0.0. This step does not upgrade the CRDs.

Reviewer: ```suggestion
# Upgrade kuberay-operator to v1.0.0. This step doesn't upgrade the CRDs.
```

$ helm upgrade kuberay-operator kuberay/kuberay-operator --version v1.0.0

> # Install a RayCluster using the v1.0.0 helm chart to verify successful upgrade

Reviewer: ```suggestion
# Install a RayCluster using the v1.0.0 helm chart to verify the success of the upgrade.
```
